flocker/acceptance/endtoend/test_leases.py

```diff
@@ -0,0 +1,238 @@
+                    binds=["{}:/data".format(dataset.path.path)],
+                    port_bindings={container_http_port: host_http_port}),
+                "ports": [container_http_port],
+                "volumes": [u"/data"]}
+            container = client.create_container(
+                "python:2.7-slim",
+                ["python", "-c", script.getContent()] + list(script_arguments),
+                **docker_arguments)
+            container_id = container["Id"]
+            containers.insert(0, container_id)
+            client.start(container=container_id)
+            self.addCleanup(client.remove_container, container_id, force=True)
+        starting_http_container = acquiring_lease.addCallback(
+            start_http_container
+        )
+
+        def write_data(ignored):
+            """
+            Make a POST request to the container, writing some data to the
+            volume.
+            """
+            data = random_name(self).encode("utf-8")
+            d = post_http_server(
+                self, cluster.nodes[0].public_address, host_http_port,
+                {"data": data}
+            )
+            d.addCallback(
+                lambda _: assert_http_server(
+                    self, cluster.nodes[0].public_address,
+                    host_http_port, expected_response=data
+                )
+            )
+            return d
+        writing_data = starting_http_container.addCallback(write_data)
+
+        def stop_container(ignored):
+            """
+            This ensures Docker hasn't got a lock on the volume that might
+            prevent it being moved separate to the lock held by the lease.
+            """
+            [container_id] = containers
+            client.stop(container_id)
+        stopping_container = writing_data.addCallback(stop_container)
+
+        def perform_operation(ignored):
+            return operation(
+                dataset_id=dataset_id, **additional_kwargs
+            )
+        performing_operation = stopping_container.addCallback(
+            perform_operation
+        )
+
+        def wait_for_unexpected_umount(ignored):
+            """
+            If a bug or error in the dataset agent is causing it to not
+            respect leases, then we expect that 10 seconds is long enough
+            for it to begin performing the requested (move or delete a
+            dataset) operation.
+            The first step of such an operation will always be to unmount
+            the filesystem, which should happen quickly since there are no
+            open files and no Docker bind mounts to the filesystem.
+            Therefore if after 10 seconds, we can restart the container and
+            successfully write some data to it, we can conclude that the
+            dataset agent has respected the lease and not attempted to
+            unmount.
+            """
+            return deferLater(reactor, 10, lambda: None)
```

Comment:
> > would it be possible to have a retry loop here that stops retrying at 10s?

> I'm not sure, I can't immediately envision how that would work. In the 10s period, we don't expect anything to happen. If we were to try restarting the container immediately in a retry-loop, we'd expect it to also succeed immediately - but that could happen even if leases weren't being properly respected, because there may not even have been time to unmount, so it wouldn't tell us anything useful to this test case.

OK, I think I understand now. You're trying to prove a negative, which is intrinsically hard. This is a good compromise.